### PR TITLE
Get temp rework take2

### PR DESCRIPTION
**Pull request by mrwgx3**

### CODE REVIEW

**Comment by mrwgx3** [comment]
Changed constructor / methods in RtcTemperature to process scaled 2's comp. integer from RTC.
Fixed old methods and added (2) new methods: AsScaledDegrees(), AsRoundedDegrees()

**Comment by Makuna** [comment]
(back from holiday, reviewing soon)

**Comment by smz** [comment]
@mrwgx3 , I've thought about what you're doing here and I'm afraid there is something wrong:

In the original API the `GetTemperature()` method returns the whole degrees and the fractional part of degree (_with wrong values for negative temperatures, as you correctly pointed out_).

In your modified version the fractional part is just the plain byte as read from the DS3231 device, so it doesn't represent the **decimal** fraction of degree.

Some user could be already using this method to print out (_or otherwise use_) those values, successfully for positive temperatures.

If your modification are accepted and the user updates his/her library, he/she would start getting some very odd results.

In a way, your modification is not "backward compatible", i.e. it breaks the API.

I think that in `GetTemperature()` you should get the correct floating point temperature (_same as you did with your PR for my library_) and then split it into its integer and fractional parts so that the results are consistent to what they were before.

**Comment by mrwgx3** [comment]
Hi @smz, and thanks for you commentary.

First off, we must remember the nature of the bug is a mis-interpretation of the meaning of the RTC temperature register contents. Lets review GetTemperature() and portions of the original RtcTemperature class:

```
RtcTemperature RtcDS3231::GetTemperature( void )
{
   .....

   _wire.requestFrom(DS3231_ADDRESS, DS3231_REG_TEMP_SIZE);

   // Old interpretation:   Signed 8-bit integer / Unsigned 8-bit scaled fraction

-  int8_t degrees = _wire.read();                // 'degrees' is SIGNED, with no scaling
-  // fraction is just the upper bits
-  // representing 1/4 of a degree
-  uint8_t fract = (_wire.read() >> 6) * 25;     // 'fract' is UNSIGNED, with SF = (256 / 64) * 25 = 100
-  return RtcTemperature(degrees, fract);        // Original 'RtcTemperature' uses degrees/fract

} //GetTemperature

class RtcTemperature
{
public:

-    RtcTemperature(int8_t degrees, uint8_t fraction) :
-        integerDegrees(degrees),
-        decimalFraction(fraction)
     {
     }

     int8_t AsWholeDegrees()
     {
-        return integerDegrees;    // This is 'degrees' supplied by GetTemperature()
     }                             // Signed and un-scaled (x1)

     uint8_t GetFractional()
     {
-        return decimalFraction;   // This is 'fract' supplied by GetTemperature()
     }                             // Unsigned and scaled (x100)
 
protected:
-    int8_t  integerDegrees;
-    uint8_t decimalFraction;

}; //end, RtcTemperature

```
When GetTemperature() is called, it returns a RtcTemperature object containing a signed/un-scaled integer temperature and a unsigned/unscaled fractional temperature. When called, AsWholeDegrees() and GetFractional() each deliver their respective temperature components unchanged:
```
a) AsWholeDegrees() returns a signed single-byte integer temperature of range -128 to +127 degC

b) GetFractional()  returns an un-signed single-byte scaled fractional temperature representing
                    100 * {0, 1/4, 1/2 and 3/4 degC} = { 0, 25, 50, and 75}
```
It's important to note that GetTemperature() does the scaling for the fractional portion.

As presented, these values would be idea for printing a float representation of signed temperature, but without needed a float conversion. If the RTC did that, it would be pretty cool, but a little odd as well. This is what prompted me to go look carefully at the temperature register description in theDS3231 datasheet:

```
Temperature Registers (11h–12h)
Temperature is represented as a 10-bit code with a resolution of 0.25°C and is accessible at
location 11h and 12h. The temperature is encoded in two’s complement format. The upper 8 bits,
the integer portion, are at location 11h and the lower 2 bits, the fractional portion, are
in the upper nibble at location 12h. For example, 00011001 01b = +25.25°C.
```

There's nothing in the above wording which rules out an 8-bit signed integer / unsigned fraction register interpretation, as inferred by GetTemperature(). Conversely, the description wording doesn't rule out 16-bit, 2's complement interpretation either, which is also easily realizable in hardware. The inability to decide conclusively prompted me to do the experiment leading up issue (#49).

Now lets look at GetTemperature() and the RtcTemperature class for corrected code:
```
RtcTemperature RtcDS3231::GetTemperature( void )
{
   .....

   _wire.requestFrom(DS3231_ADDRESS, DS3231_REG_TEMP_SIZE);

   // Correct interpretation:  Signed 16-bit, 2's complement integer
+  int8_t  mstmp = _wire.read();                 // MS byte, scaled 2's comp. temperature
+  uint8_t lstmp = _wire.read();                 // LS byte, ............................
+  return RtcTemperature( mstmp, lstmp );        // New 'RtcTemperature' uses mstmp/lstmp 

} //GetTemperature

class RtcTemperature
{
public:

+    RtcTemperature( int8_t msdegC, uint8_t lsdegC ) :
+        mstemp( msdegC ),
+        lstemp( lsdegC & 0xC0 )  // Bind to RTC resolution (only upper (2) bits enter computation)
     {
     }

+    // Scaled integer temperature (x 256)
+    int16_t AsScaledDegrees()
+    {
+        return(  (mstemp << 8) + lstemp  );  // Concatentate raw temp. regs 
+    }

+    // Rounded integer temperature (x 1)
+    int8_t AsRoundedDegrees()
+    {
+        // Equivalent float function:
+        //   SF ( sign(x)    * floor(  abs(       x ) + 0.5    ) ) / SF, with
+        // =    ( sign(SF x) * floor(  abs(    SF x ) + 0.5 SF ) ) / SF  SF = 256
+        // =      sign(x)    * floor( sign(x)( SF x ) + 0.5 SF )   / SF
+        //
+        int8_t  sgnT = ( (mstemp < 0) ? -1 : 1 );
+        return( sgnT * ((sgnT * AsScaledDegrees() + 0x80) >> 8) );
+    }

     // Float temperature (x 1)
     float AsFloat()
     {
+        return ( (float)AsScaledDegrees() / 256.0f );
     }

+    // Integer portion, temperature (x 1), useful for printing
     int8_t AsWholeDegrees()
     {
+        int8_t  sgnT = ( (mstemp < 0) ? -1 : 1 );
+        return( sgnT * ((sgnT * AsScaledDegrees()) >> 8) ); // (x 1)
     }
 
+    // Fractional portion, temperature (x100), useful for printing
+    // Returns (0, 25, 50, or 75)
     uint8_t GetFractional()
     {
+        int8_t  sgnT = ( (mstemp < 0) ? -1 : 1 );
+        return(   25 * ((uint8_t)(sgnT * lstemp) >> 6)  );  // (x 100)
     }

protected:
-    int8_t  mstemp;
-    uint8_t lstemp;

}; //end, RtcTemperature
```

When GetTemperature() is called, it now returns a RtcTemperature object containing a 2's complement (i.e, signed) 16-bit integer; the same information content as the original, but in a different format.

Note also that the methods in RtcTemperature now do all the heavy lifting; GetTemperature() just passes the raw temperature registers along.

Now that we've reviewed code, it would be a good time for a numerical example. First off, lets define our primary variables as the ones returned by GetTemperature():
```
a) degrees / fract   for the original  version
b) mstemp  / lstemp  for the corrected version
```
```
Say my RTC is at a temperature of 23.25 degC, which scales to 256 x 23.25 = 5952, or 0x1740.

Here 'mstemp' and 'degrees' both equal 0x17 = 23, and
     'lstemp' = 0x40 = 64, and 'fract' = (lstemp / 64) * 25 = 25.  All well and good, both original and new

Now say my RTC is at a negative temperature of -23.25 degC, which scales to 256 * -23.25 = -5952, or 0xE8C0.

Here 'mstemp' and 'degrees' now equal 0xE8 = -24,                       which is not -23, and
     'lstemp' = 0xC0 = 192, and 'fract' becomes (lstmp / 64) * 25 = 75, which is not  25
```
As you can see, splitting apart a 16-bit, 2's complement integer breaks the original interpretation of 'signed unscaled / unsigned scale fractional' for negative temperatures. This is the bug quantified.

Now lets plug our numbers into the new methods of RtcTemperature:
```
AsScaledDegrees() recombines 'mstemp' and 'lstemp' back into a single integer:

returned integer equals:  (mstemp << 8) + lstemp
                          (  0xE8 << 8) + 0xC0
                               0xE800   + 0xC0
                          0xE8C0 = -5952, our orignal signed integer

Both AsWholeDegrees() and GetFractional()  now use interm
'sgnT' to extract the sign of the 2's complement integer:

sgnT = ( (mstemp < 0) ? -1 : 1 );
     = ( (-24 < 0) : -1 : 1 );
     = -1

AsWholeDegrees() returns:

   sgnT * ((sgnT * AsScaledDegrees()) >> 8) 
   (-1) * ((-1 * -5952) >> 8 )
   (-1) * ( 0x1740 >> 8 )
   (-1) * ( 0x17, or 23 )
   -23    No problems now

GetFractional() returns:

    25 * ((uint8_t)(sgnT * lstemp) >> 6)
    25 * ((uint8_t)( -1 * 0xC0 ) >> 6 )
    25 * ((uint8_t)( -1 * -64  ) >> 6 )
    25 * ((uint8_t)( 64 >> 6 )
    25 * ((uint8_t)( 1 )
    25   No problem here either; do you recognize the scaling used in the original GetTemperature()?
```
As you can see, both AsWholeDegrees() are GetFractional() now process negative temperatures correctly; they're fixed! They are of great use for printing, but not much use computationally-wise.

Concerning method AsFloat(), I hope you can see now that its exactly the same endian-independent function I tested for your library. As for method AsScaledDegrees(), it is integral to most the the methods in the new RtcTemperature. I decided to keep it public, as it:
```
a) Is generally useful, computationally-wise.
b) Can be used to reconstruct the original RTC temperature registers.
```

What is most useful, integer-wise, is a rounded version of temperature, which is provided by method AsRoundedDegrees() (you can verify it as an exercise).

Hope this addresses you concerns!

**Comment by smz** [comment]
@mrwgx3 the problem (_B/C breach_) would appear if someone have a sketch like this (_simplified_):
```C++
RtcDS3231<TwoWire> Rtc(Wire);

RtcTemperature myTemperature;

myTemperature = Rtc.GetTemperature();

Serial.print("Temperature is: ");
Serial.print(myTemperature.integerDegrees);
Serial.print(".");
Serial.print(myTemperature.decimalFraction);
```
The above sketch (_forgetting about possible mistakes or typos!_) is perfectly legit, but wouldn't work anyn
more with your mods, I think.

**Comment by smz** [comment]
... wait... I see that `integerDegrees` and `decimalFraction` are **protected** members of the RtcTemperature class, so maybe my hypothetical sketch wouldn't had worked in the first place and you can freely rename/re-purpose them...

**Comment by mrwgx3** [comment]
Ah, I think we're getting to the bottom of your confusion.

If ```integerDegrees``` and  ```decimalFraction``` were declared 'public', your code would print out their values.  Being declared 'protected', however, is pretty much equivalent to 'private' here, hence you should get an error when compiling.

If you changed your code slightly, it would work, but you'd probably not like the result formatting:
```
myTemperature = Rtc.GetTemperature();

Serial.print("Temperature is: ");
Serial.print( myTemperature.AsWholeDegrees() );
Serial.print(".");
Serial.print( myTemperature.GetFractional() );
```
The methods AsWholeDegrees() and GetFractional() serve as getters for ```integerDegrees``` and  ```decimalFraction```, repectively.  You can read them, but can't modify them. For the corrected code, these methods do much more.

Note also AsWholeDegrees() and GetFractional() only returned an 8-bit signed/unsigned integer,  Serial.print() can't do much with them. As for your comment,
```
The above sketch (forgetting about possible mistakes or typos!) is perfectly legit,
 but wouldn't work anymore with your mods, I think.
```
the original or corrected code would perform just the same, because the variable types used haven't changed.

Now I shall clarify what I mean by 'great for printing'

To write a general Print class can be a very convoluted affair, as one needs consider all the nuances of the format string which need to be parsed and interpreted. Furthermore, the amount of code generated can be horrendous, and burn up most of your embedded processor's code space, esp. if floats are involved.  Printing integers and strings is much simpler, and takes much less code space.

More often than not, the Print class you get has been stripped of its ability to print floats. This certainly seems true for the ESP8266  Arduino code (may just be compiler flags not setup), and I suspect is true for most supported Arduino boards in general.

For example, the follow 'printf' call doesn't work on my ESP8266:
```
    float t1 = -23.25;
    Serial.printf( "%5.2f\n", t1 );     // Just outputs '%.2f' to my serial monitor
```
I can use ``Serial.print( )``` to print a float, but it doesn't provide tight control over formatting; too general a tool, not enough finesse. If I can extract the integer and fractional components of my float, I can do a good job using 'Serial.printf()':
```
    int8_t  d1 = myTemperature.AsWholeDegrees();   // Assume -23 is returned
    uint8_t f1 = myTemperature.AsWholeDegrees();   // Assume   0 is returned
    Serial.printf( "%3d.%02u\n", d1, f1 );        // Outputs '-23.00'

Note that I didn't put my function calls directly inside the printf() function call;, as
I remember reading that the current 'printf()' can handle that 
```
I got my output looking like a float, but I only had to use integers to do it. Great for printing!

**Comment by smz** [comment]
Hello @mrwgx3 and thank-you for your exhaustive explanation!

I understand the limits of the Arduino Print class and the rationale for having the AsWholeDegrees(), GetFractional() and the new AsRoundedDegrees() methods.

My example sketch was obviously wrong (_conceptually and syntactically_). 

I'm an old dinosaur raised on Fortran, Pascal, PL/1 and derivatives, C (K&R), plus a little bit of Lisp, Basic and PHP here and there:  my procedural imprinted mind still struggles absorbing the beauties and intricacies of OOP, so I hope you will forgive me if I bring out a similar but different example.

TLDR;

**Couldn't be the case that there, is floating around in the Arduino ecosystem, a derivative class of the RtcTemperature class having access to the original** `integerDegrees` **and** `decimalFraction` **attributes?**

**If** (_big if..._) this **can be** the case, then my original objection still holds and in that case I think a safer implementation of your work could be:

1) do not rename the `integerDegrees` and `decimalFraction` attributes

2) in GetTemperature() do something like this (_to be checked / simplified / optimized..._):

```C++
int8_t degrees = _wire.read();
uint8_t fract = _wire.read()
float f_degrees = (int16_t) ((degrees << 8) | fract ) / 256.0f;
fract = (f_degrees - (int8_t) (f_degrees)) * 100;
return RtcTemperature(degrees, fract);
```

From there on, everything will work as before and of course you can always add the (_useful_) AsRoundedDegrees() method.

There will obviously be a small computational inefficiency (_going back and forth with the float value_), but I think that would be really negligible...

**Comment by smz** [comment]
... also, I think that while the `integerDegrees` and `decimalFraction` attributes have an universal meaning, that could apply to value returned by different hardware beside the DS3231 (_that could possibly be supported in the future_), your `mstemp` and `lstemp` attributes are too tightly coupled to the specific implementation of the DS3231 hardware.

**Comment by mrwgx3** [comment]
Hi @smz
Passing through at the moment, your last comment noted. Will get back to you  later today

**Comment by Makuna** [comment]
In reality what you are defining is a more accurate fixed point math object with your implementation; which I like.  Please see review for specific change requests; but below are some thoughts for discussion.

Something to note about float on AVR; the code is large and slow.  So some people will want to avoid it completely if they build for ATTINY devices which have very small flash sizes.  The compiler will remove methods that aren't used, so have it converted to float in a method is ok; but storing it as float is not.  

Further, the intent of exposing the integer and fractional parts was for printing purposes and avoiding the float code in the printing code also.

File: src/RtcTemperature.h [hunk 1]
```diff
@@ -7,32 +7,54 @@
...
-        integerDegrees(degrees),
-        decimalFraction(fraction)
+    RtcTemperature( int8_t msdegC, uint8_t lsdegC ) :
+        mstemp( msdegC ),
+        lstemp( lsdegC & 0xC0 )  // Bind to RTC resolution
     {
     }
 
+    // Scaled integer temperature (x 256)
+    int16_t AsScaledDegrees()
```
Comment (Makuna) [inline src/RtcTemperature.h]: I don't like the name of this just due to the fact that its basically the raw "8.8 fixed point" math item.  
"AsFixedPointDegrees" would be my suggestion.

File: src/RtcTemperature.h [hunk 2]
```diff
@@ -7,32 +7,54 @@
...
-        return degrees;
+        return ( (float)AsScaledDegrees() / 256.0f );
     }
 
+    // Integer portion, temperature (x 1), useful for printing
     int8_t AsWholeDegrees()
     {
-        return integerDegrees;
+        int8_t  sgnT = ( (mstemp < 0) ? -1 : 1 );
+        return( sgnT * ((sgnT * AsScaledDegrees()) >> 8) ); // (x 1)
```
Comment (Makuna) [inline src/RtcTemperature.h]: Why do the extra work within AsScaledDegrees to just undo it here.  Just return what you are currently calling mstemp.

File: src/RtcTemperature.h [hunk 3]
```diff
@@ -7,32 +7,54 @@
...
     {
-        return decimalFraction;
+        int8_t  sgnT = ( (mstemp < 0) ? -1 : 1 );
+        return(   25 * ((uint8_t)(sgnT * lstemp) >> 6)  );  // (x 100)
     }
 
 protected:
-    int8_t integerDegrees;
-    uint8_t decimalFraction;
+    int8_t   mstemp;
```
[the remark this answers is not in the document]
Comment (mrwgx3) [reply src/RtcTemperature.h]:
```
     int8_t AsWholeDegrees()
     {
-        return integerDegrees;
+        int8_t  sgnT = ( (mstemp < 0) ? -1 : 1 );
+        return( sgnT * ((sgnT * AsScaledDegrees()) >> 8) ); // (x 1)

@Makuna
Why do the extra work within AsScaledDegrees to just undo it here.
Just return what you are currently calling mstemp.
```
I see by your question I still haven't communicated the fundamental nature of the bug.

Before answering your other questions, let's start with an allegory:

Say you just got the preliminary spec. on Maxim's newest RTC, the DS3231HE, the RTC designed for HELL,
and asked me to code the part. As Hell's denizens wished to precisely predict when Hell freezes over while
it's happening (a New Year's Eve BallDrop kind of thing, only more so), Maxim's designers gave them the
best 2's complement fixed point word width and alignment they could come up with:
``` 
      |         r11h           |      DP     r12h         |
Bit:   15 14 13 12 11 10  9  8   7  6  .  5  4  3  2  1  0  -1 -2 -3
        s  i  i  i  i  i  i  i   i  i  .  f  f  f  f  f  f   0  0  0
```
Only (2) extra bits were needed for the integer portion, because even when Hell
freezes over in its biggest way, it still only approaches absolute zero, or -273.15 degC.
Due to the extra integer bits, however, the decimal point is no longer aligned at a byte boundary.

As it takes (6) right-shifts to register the decimal point (DP) to the right of the 0th bit,
the overall word scaling equals 64. Given (6) fractional bits, our fractional resolution
is 1/64th of a degree.

To write a new AsWholeDegrees() to accommodate the off-boundary alignment, one must:
```
     a) Concatentate the temperature registers into a single 16-bit signed integer
     b) Record the sign of the 16-bit result
     c) Find the absolute value of the concantenation
     d) Align the LSbit of the integer portion with the 0th bit of the word
     e) Restore the sign of the result, then return it
```
Writing a new GetFractional() is just as easy:
```
   a-c) Do steps (a-c) of the prototype AsWholeDegrees() function.
     d) Align the LSbit of the fractional portion with the 0th bit of the word 
     e) Extract the fractional part with a mask, multiply by an appropriate
        scale-factor, then return the result.
```
By noting that  Abs(x) = sign(x) * x, I can write these routines very compactly.
```
// For the DS3231HE, RSI_GETFRAC = 6
// Function returns -512 through 511
//
int16_t Proto_AsWholeDegrees( int8_t r11h, uint8_t r12h )
{
   int16_t wordTemp = (int16_t)( (r11h << 8) + r12h ); // (a)
   int8_t  sgnT = ( wordTemp < 0 ) : -1 : 1;           // (b)
   return  sgnT * ((sgnT * wordTemp) >> RSI_GETFRAC);  // (c-e)

} //Proto_AsWholeDegrees

// For the DS3231HE,
//    SF_GETFRAC = 15625, MASK_GETFRAC = 0x003F, and
//    RSF_GETFRAC = 0, as fractional part is already aligned
//    Function returns  { 0, 15625 through 984375 }
//
uint32_t Proto_GetFractional( int8_t r11h, uint8_t r12h )
{
   int16_t wordTemp = (int16_t)( (r11h << 8) + r12h );     // (a)
   int8_t  sgnT = ( wordTemp < 0 ) : -1 : 1;               // (b)
   return SF_GETFRAC *
      ( MASK_FRAC & ((sgnT * wordTemp) >> RSF_GETFRAC) );  // (c-e)

} //Proto_GetFractional
```
As I'm finishing up the above code, however, Maxim sends a final project update quoting Hell's CEO: 

"Given general human nature, Hell isn't isn't likely to freeze over anytime soon. Furthermore, with the
recent upturn in bizarre human behavior, there is no lack of "cold days", either. The standard part will
do fine, and the DS3231HE project is shelved until further notice."

No worries, just a final tweek of the constants:
```
The original fixed point word/alignment is:

      |         r11h          | DP |         r12h         |
Bit:   15 14 13 12 11 10  9  8   .  7  6  5  4  3  2  1  0  -1 -2 -3
        s  i  i  i  i  i  i  i   .  f  f  0  0  0  0  0  0
and

RSI_GETFRAC  = 8       to align the LSbit of the integer    part to bit 0
                       (overall word scale-factor is 256)
RSF_GETFRAC  = 6       to align the LSbit of the fractional part to bit 0
MASK_GETFRAC = 0x0003  mask to extract the fractional part
SF_GETFRAC   = 25      scale to decimal

GetFractional() now returns {0, 25, 50, and 75 }
```
But wait you say, "With the decimal point now falling on a byte boundary, along with the greatly reduced resolution, can't one simplify the code such that AsWholeDegrees() and GetFractional() are only functions
of 'r11h' and 'r12h', respectively?
```
Alas, no, due to the "1's complement + 1" nature of 2's complement (signed) arithmetic:

   a) Knowing when to negate both integer and fractional portions always needs sign test of the integer part.

   b) Adding (1) to the negated integer portion is only required when the original fractional part equals zero,
      hence requires a test of the fractional part.

It's inescapable, both temperature registers are needed for AsWholeDegrees() or GetFractional().
```
#### This is what I assumed you meant by your question, Makuna. ####  
No work is being undone, and 'mstemp' aka 'r11h' can't do the job alone.

Back to your questions, @smz. Noting your concerns about derived classes and coding for other RTC's,
I searched the web for:
```
   a) Any Arduino libraries written for the DS3231 RTC
   b) Any comparable parts, i.e., with an onboard, temperature compensated crystal oscillator
```
Regarding libraries, of the dozen or so that I examined, I only found (1) other library that looks similar to Makuna's. But as in your library, they tossed the RtcTemperature class and re-coded GetTemperature() to return a float.

Of the remaining libraries, around 40% used the 16-bit integer interpretation, 30% picked the deg/frac approach, and one even merged the two concepts together.

Regarding other RTC's, I only found (1) other device which returned a temperature, Intersil's ISL12022 RTC.
https://www.intersil.com/content/dam/Intersil/documents/isl1/isl12022.pdf

Their temperature register description (pg 21) is even more confusing than Maxim's. Their design is slightly more forward thinking as they've anticipated Hell freezing over; temperatures are reported in degrees Kelvin.

#### It would seem that the DS3231 is a unique part, and Makuna's library a unique approach. ####
I wouldn't worry too much about breaking derived classes.

Back to your remaining comments, @Makuna.

After finishing my research, I've come to the conclusion that the person who wrote the original temperature
register description did not understand the 2's complement nature of the registers. Consider instead:
```
Temperature Registers (11h–12h)
Temperature is represented as a 10-bit code with a resolution of 1/4th °C and is accessable as
a 16-bit, signed integer at locations 11h and 12h. For example, at +/- 25.25°C, concatenated
registers <R11h:R12h> = 256 * (+/- 25+(1/4)) = +/- 6464, or 1940h / E6C0h.
```
#### The above description is correct, as determined by code verified through experiment. ####

For our allegorical 'HE' part, the description is just as succinct:
```
Temperature is represented as a 16-bit code with a resolution of 1/64th °C and is accessable as
a 16-bit, signed integer at locations 11h and 12h. For example, at +/- 25.27°C, concatenated
registers <R11h:R12h> ~= 64 * (+/- 25+(17.28/64)) = +/- 1617, or 0651h / F9AFh.
```
No need to mention integer/fraction alignments to confuse things.

If you'd read this description first, what primary variables would your have used in GetTemperature()?
Passed to and saved in RtcTemperature? Where would you have put your maths?
```
 protected:
-    int8_t integerDegrees;
-    uint8_t decimalFraction;
+    int8_t   mstemp;
+    uint8_t  lstemp;

In reality what you are defining is a more accurate fixed point math object with your implementation.

"integer" and "fractional" are still better variable names for these, not most significant and least significant.

BTW: avoid the name "temp" for variables as it really isn't descriptive to the object or algorithm in most cases, and specifically in this usage.
```
Regarding 'fixed point math object'...

No, I'm just using the fixed point math object provided by the DS3231.

Regarding 'integer" and 'fractional' vs 'mstemp' and 'lstemp'...

As I hope was shown by the allegory, "integer" and "fractional" in a 2's complement context are really describing
bit-field alignments. As demonstrated by the re-written register description, bit-fields are not required
to understand the DS3231 temperature register contents.

'mstemp' and 'lstemp' just transport these bit-fields, but I agree, a more descriptive primary variable could be defined.

Regarding the BTW...
'temp' is descriptive in the context of temperature, but could construed as 'temporary'.

```
+    // Scaled integer temperature (x 256)
+    int16_t AsScaledDegrees()

I don't like the name of this just due to the fact that its basically the raw "8.8 fixed point" math item.
"AsFixedPointDegrees" would be my suggestion.
```
My first inclination was to name it simply 'AsInteger()' in analogy to 'AsFloat()', but it has a non-unity
scale-factor. "AsFixedPointDegrees()" works for me.


As a new primary variable(s), I would propose using:
```
private:
int16_t  fxdpntTemperature;
int8_t   sgnT;               // Sign of 'fxdpntTemperature'
```
GetTemperature() still transfers register values:
```
RtcTemperature GetTemperature ( void )
{
   ....
   _wire.requestFrom(DS3231_ADDRESS, DS3231_REG_TEMP_SIZE);
   int8_t  r11h = _wire.read();                  // MS byte, signed temperature
   return RtcTemperature( r11h, _wire.read() );  // LS byte is r12h
}
```
The constructor creates the primary variable:
```
RtcTemperature( int8_t ubRegT, uint8_t lbRegT )
{
   fxdpntTemperature =  (int16_t)( (ubRegT << 8) | (lbRegT & 0xC0) );
   sgnT = ( (fxdpntTemperature < 0) ? -1 : 1 );  // Sign of 'fxdpntTemperature' 
}
```
Method 'AsScaledDegrees()' becomes 'AsFixedPointDegrees()'
```
//  Used either to...
//  a) As an interger alternative for temperature calculations
//     equals 256 * (temperature in degC)
//  b) Reconstruct the original temperature registers
//
int16_t  AsFixedPointDegrees ( void )
{
   return fixedpntTemperature;
}
```
AsScaledDegrees() would be then be replaced by 'fixedpntTemperature' in all other methods.

No direct access to primary var's, and maths are all in one place. What do you think?

[the remark this answers is not in the document]
Comment (Makuna) [reply src/RtcTemperature.h]: Why do you have a separate sign member and a int16_t member above?  

Instead of "T", "temp", or even "temperature", how about  just replace temp with "degrees" or even "degreesC".  

leave off describing the type from the variable name (fixedpnt) and just point a comment next to the variable.

[the remark this answers is not in the document]
Comment (mrwgx3) [reply src/RtcTemperature.h]: You're a tough customer, Makuna!

Regarding the 'sgnT' variable, I have (4) instances of
```
sgnT = ( (fxdpntTemperature < 0) ? -1 : 1 );  // Sign of 'fxdpntTemperature'
``` 
in RtcTemperature. By adding 'sgnT' to private vars and initializing it in the constructor, I can reduce the number of instances to (1). Have restored all original instances.

Have changed all instances of 'fixedpntTemperature' to 'degreeC'

[the remark this answers is not in the document]
Comment (mrwgx3) [reply src/RtcTemperature.h]: All the changes I've made in response to your comments I've placed in a branch of my forked repository called ['getTemp_rework_take3'][rework3]. It is as yet un-merged with branch  ['getTemp_rework_take2'][rework2], the branch for this pull request.

I've invited you to be a collaborator for this repo, so you can make your desired changes, then merge the branches when you're satisfied.

While testing, a discovered a conceptual error regarding the usage of 'AsWholeDegrees()' and 'GetFractional(). Consider the temperatures -0.25, -0.50, and -0.75 degC.  In each case, 'AsWholeDegrees()' returns zero as there's no way to define a 'negative zero'. To get around this problem, I defined a new method called ['IsMinusSignNeeded()'][minus_test], which is used in the updated ['DS3231_Temperature.ino'][minus_print] sketch. BTW, it appears that 'printf_P()' is non-functional.

[rework2]: https://github.com/mrwgx3/Rtc/blob/getTemp_rework_take2/src/RtcTemperature.h "1st cut"
[rework3]: https://github.com/mrwgx3/Rtc/blob/getTemp_rework_take3/src/RtcTemperature.h "Latest changes"
[minus_test]: https://github.com/mrwgx3/Rtc/blob/getTemp_rework_take3/src/RtcTemperature.h#L59#L64 "Leading minus test"
[minus_print]:https://github.com/mrwgx3/Rtc/blob/new_example_DS3231_Temperature/examples/DS3231_Temperature/DS3231_Temperature.ino#L169#L182 "How 'IsMinusSignNeeded()' is used"

[the remark this answers is not in the document]
Comment (Makuna) [reply src/RtcTemperature.h]: Ahh, ignore my comment in the sample code, now I understand why you have sign separate. Let me think on this a bit, it sort of brings the original reasons for these into question and maybe there is a better approach.

[the remark this answers is not in the document]
Comment (mrwgx3) [reply src/RtcTemperature.h]: No worries. Given the lack of Arduino float formatting, I think the idea is a needed one. Question is, how do want to deliver the information?  The information content for printing is a triplet, { sign, whole degrees, fractional degrees}. Perhaps if you return a fixed-width text string with leading spaces (one which does NOT use 'printf', but  BCD routines like the RTC.cpp section), most people would be satisfied with that.  You then could also make the triplet available for those who want to construct custom formats using 'printf' in a manner similar to the one used in the 'DS3231_Temperature.ino'  code.
